### DISCUSSION TITLE
About ungoogled-chromium-android and Bromite

### DISCUSSION THREAD

**Discussion by wchen342** [bromite/bromite#957] [General]
*First a disclosure: I now work part time for Brave. However, I have explicitly asked for permissions to continue working on open source projects in the contract and they shall not interfere except for intellectual property/copyright problems.*

I have been thinking about merging the existing `ungoogled-chromium-android` patches into Bromite as you think fit, and then start working on something else. There are two reasons:

 - A large part of the functionalities of `ungoogled-chromium-android` already exists in Bromite, including removal of google libraries. If you look at [this](https://github.com/ungoogled-software/ungoogled-chromium-android#differences-from-ungoogled-chromium) you will see a large portion of the added functionalities are borrowed from Bromite.
 - I feel there is not a strong need for an Android version chromium that only removes google library. If that is the need, it has already been fulfilled by Bromite, and if someone actually want stronger privacy protection, Bromite does that better than uc-android.
 
Sometimes I actually think we do not need an Android version of `ungoogled-chromium`, but rather a desktop version of Bromite. Not sure whether you have thinked of that before or not.

Also about the extensions, if you already have ad blockers inside Bromite I think the value of extensions diminishes a lot, since a large part of the needs for extensions is because people want an ad blocker. I like extensions because it is a more general solution, but I feel that it doesn't fit well with Bromite because extensions adds a lot to things like fingerprint. Maybe I can give some advice on the in-development user scripting patch though.

A side note: I found [something](https://www.reddit.com/r/androidapps/comments/lpelra/replaced_bromite_with_ungoogled_chromium_extension) interesting on Reddit. Someone seems to be saying Bromite is slower. I guess it is because of the ad blocker?

About the logics behind why I didn't use Bromite before, I have stated the main reasons in the other discussion in ungoogled-chromium, but I would like to add a little more context. So from what I can see, Bromite has a lot of issues extended from its modifications, and I guess it has something to do with the patches not being thoroughly tested as you said there are not enough resources. That was the first impression for me which stopped me from using it, but I guess it is actually stable enough, just not "as stable" as the vanilla chromium stable version.

The biggest complaint of mine about Bromite is actually its probably most welcomed functionality, the ad blocker (and now the user script too). This is only my personal view but I think it is a pretty big change, yet it is not tested properly. That means the change can make the whole browser prone to attacks, but I am not really an expert in security so I cannot comment further on this. I just feel like this needs to be addressed somehow or at least be communicated to the users, as I kept warning the extensions are highly experimental.

**Comment by ghost** [comment]
@wchen342  Congrats on being hired by brave 🎉 
(Looking forward to alpha brave builds with extension support 😜)
Wish you good luck !
As I see uc and bromite are two different things bromite being a privacy only concerned project while uc being ungoogled and some quality of life features, the recent extension update is awesome!! UC is a drop in replacement for Chrome,which even normal people can use without much thought.
So please don't discontinue uc

**Reply by wchen342** [reply]
I will not immediately discontinue anything, rather turn into a "maintainance mode" which I keep the version level with stable chromium release but stop adding new features for now.
But I do wonder, how much people think Bromite is from a drop-in replacement of chromium? I think as long as you disable the ad blocker they shall be pretty similar, or do I miss something?

**Reply by chromer030** [reply]
Bromite has some unique features like Proxy and Custom-User agent support while chromium is bare metal.

**Reply by wchen342** [reply]
The problem is people are asking me to add those to ungoogled-chromium too, so if I keep taking the requests it will eventually become the same as Bromite, just slower.

I feel that I need to rethink what is really needed instead of blindly following requests and borrowing patches from Bromite. Also I was kind out burnt out after finished the extensions menu thing so I need to take a break.

**Reply by csagan5** [reply]
> chromium is bare metal.

Chromium still has a lot of integrations that "call home"

> The problem is people are asking me to add those to ungoogled-chromium too, so if I keep taking the requests

You do not have to :) if you have 1000 users, only 2-3 will complain and come to ask for a feature or a change. If you lose 100 because of that change, maybe 1 or 0 will come back to tell you. And so on.

I always suggest to focus on project goals and mission, not on "feature parity" or any other nonsense.

> I feel that I need to rethink what is really needed instead of blindly following requests and borrowing patches from Bromite.

If the browser is called ungoogled-chromium, then it should follow UC's project goals as close as possible. If those are achieved it's basically "mission accomplished" and one has to acknowledge that.

**Reply by wchen342** [reply]
> If those are achieved it's basically "mission accomplished" and one has to acknowledge that.

Good point, I think I will just do that. The main goal *is* accomplished a long time ago so I will just keep it in that state.

**Comment by uazo** [comment]
we don't know each other but of course congratulations from me too,
as a member of the category of programmers, it is always nice to know when the skill of the individual is recognized.

> Maybe I can give some advice on the in-development user scripting patch though.

of course you are welcome!

> I think as long as you disable the ad blocker they shall be pretty similar, or do I miss something?

in my opinion, bromite is far from chromium of 150 patchs focused in privacy and respect for the user..

> That means the change can make the whole browser prone to attacks

there are so many things still to be done in bromite, first of all the reactivation of the tests. do we work together for that?

**Reply by wchen342** [reply]
> all the reactivation of the tests. do we work together for that?

That is one thing I would like to do, but I have limited time so I have to give up some thing in exchange for another. I feel like I would like to improve one single project (which is Bromite) rather than developing a separate but incomplete project parallelly.

So yeah, if you plan to work on that, open a thread somewhere so I know where to start.

**Comment by csagan5** [comment]
> _First a disclosure: I now work part time for Brave. However, I have explicitly asked for permissions to continue working on open source projects in the contract and they shall not interfere except for intellectual property/copyright problems._

Congratulations for your new engagement; however, please note that "except for intellectual property/copyright problems" refers to the totality of problems that can arise :) please always make sure that you have permission to work on projects under an open source license.

> * If you look at [this](https://github.com/ungoogled-software/ungoogled-chromium-android#differences-from-ungoogled-chromium)

The description of Bromite could probably be improved to follow what you did there.

> * I feel there is not a strong need for an Android version chromium that only removes google library. If that is the need, it has already been fulfilled by Bromite, and if someone actually want stronger privacy protection, Bromite does that better than uc-android.

I believe that diversity in open source is a strength; but if you do not see many characterising differences between the two then I am happy of your choice, Bromite will surely benefit from it.



> Sometimes I actually think we do not need an Android version of `ungoogled-chromium`, but rather a desktop version of Bromite. Not sure whether you have thinked of that before or not.

Yes, I tried to build it in the past. However I refer to ungoogled-chromium as the closest browser that is like "Bromite on desktop", so I have never spent too much time on this. This deserves its own separate discussion, but what do you think would be the differences between Bromite on desktop and ungoogled-chromium? A set of stricter defaults and some patches that are currently missing from UC?

> Also about the extensions, if you already have ad blockers inside Bromite I think the value of extensions diminishes a lot, since a large part of the needs for extensions is because people want an ad blocker. I like extensions because it is a more general solution, but I feel that it doesn't fit well with Bromite because extensions adds a lot to things like fingerprint. Maybe I can give some advice on the in-development user scripting patch though.

The biggest problems I see with extensions are:
* if upstream does not support them on Android, then all the support has to be done downstream
* the necessary UI work

I do not know how well Android Chromium forks do on the above, these days, but I have never considered it in scope for Bromite; it deserves its own separate project.

I am sure @uazo will be happy to hear from you on the user scripts patch; we are having a discussion [on the PR](https://github.com/bromite/bromite/pull/857).

> A side note: I found [something](https://www.reddit.com/r/androidapps/comments/lpelra/replaced_bromite_with_ungoogled_chromium_extension) interesting on Reddit. Someone seems to be saying Bromite is slower. I guess it is because of the ad blocker?

As for all bug reports, first one has to test with the corresponding Chromium version. Most of these reports are due to the user starting with a fresh install.

> About the logics behind why I didn't use Bromite before, I have stated the main reasons in the other discussion in ungoogled-chromium, but I would like to add a little more context. So from what I can see, Bromite has a lot of issues extended from its modifications, and I guess it has something to do with the patches not being thoroughly tested as you said there are not enough resources. That was the first impression for me which stopped me from using it, but I guess it is actually stable enough, just not "as stable" as the vanilla chromium stable version.

It depends from your definition of stable: occasional problems, yes. Browser crashing all the time: no, never happened because of Bromite's patches. To memory the biggest issues related to crashes have been due to the ARM64 video decoding, which is related to some upstream/Android issue that unfortunately requires using specific hardware to find out (tests would not help there either).

I have accepted more patches in the last year, and there have been problems related to those patches; we are still trying to strike a balance between feature development and stability.

> The biggest complaint of mine about Bromite is actually its probably most welcomed functionality, the ad blocker

I do not judge the ad blocker to be a big change, the engine is the vanilla Chromium subresource filter; the patch only introduces functionality to download the filters. These filters are downloaded via the extensions servers in the case of Chromium, and via HTTPS in the case of Bromite; I cannot see any security compromise there. Please point out any weakness you can find; if they exist, they can be solved.

> and now the user script too). This is only my personal view but I think it is a pretty big change, yet it is not tested properly. That means the change can make the whole browser prone to attacks, but I am not really an expert in security so I cannot comment further on this. I just feel like this needs to be addressed somehow or at least be communicated to the users, as I kept warning the extensions are highly experimental.

I have similar concerns and in general Bromite must stay away from features that are not related to privacy or security; user scripts has more to do with user freedom, and thus can help with privacy. The idea of user scripts is that if the patch is merged security must not be compromised; feel free to participate in the discussion there.

In general if your phone is compromised and:
* some other app modifies your browser proxy settings, that is **not** a Bromite security issue
* some other app abuses root powers and installs a custom nefarious user script, that is **not** a Bromite security issue as well
* the user installs a malicious user script, that is **not** a Bromite security issue (it can be an UI issue though, see discussion on the PR)

All Android apps work in sandboxes guaranteed by the OS, thus they need to trust the OS; rooted phones come with their own set of different risks that the user must be aware of.

But I risk going off-topic; the discussion for user scripts can continue [there](https://github.com/bromite/bromite/pull/857).

**Reply by wchen342** [reply]
> always make sure that you have permission to work on projects under an open source license

The expression in my previous statement is not exactly accurate, but I have consulted with a lawyer before signing and basically as long as I do not use overlapping codes there is no problem.

> but what do you think would be the differences between Bromite on desktop and ungoogled-chromium?

`ungoogled-chromium` as Eloston has put it, is focusing on one thing in particular, which is removing Google dependencies. That is good for someone who just want a daily drive with Google services removed, but for someone who wants more privacy, it doesn't appear to be enough. On desktop since we can use extensions, ad blocker and user script will not be that important, so the benefit of Bromite will most likely come from all the other privacy-oriented patches.

For me, when I wanted something and it doesn't exist then it is time to think of creating something new. I has always wanted something in the chromium domain that is similar to TBB. That is the reason I brought up the Firefox Tor Uplift project, as I think if something similar can be done with chromium it will be very beneficial. One simple example is the canvas, in Firefox you have [RFP with a single value](https://github.com/arkenfox/user.js/issues/767#issuecomment-515780193) across users. I probably have some over-trusting in Tor Project, I don't really know, but from my past experience and research while being behind the [Great Firewall](https://en.wikipedia.org/wiki/Great_Firewall) for years I believe Tor Project is more trustworthy than a lot of other projects out there.

> if upstream does not support them on Android, then all the support has to be done downstream

Yes, that is actually the biggest problem after I have tried. The amount of work needed is huge. Almost half of the problems comes from a single designing choice though: the tab models. They use a different set of models between desktop and android, which makes the whole API system broken.

I am actually not sure how valuable it is to maintain the extension part. It takes a huge effort and from what I can see the interest is not that big. It will be good if there are more contributors but I guess it is hard to find someone who knows the chromium code base to start with.

> All Android apps work in sandboxes guaranteed by the OS

This is more like a general question, but I was actually thinking of something the other way around: what is the chance some code executions can cause a sandbox escape? When I was talking about the risks I was thinking about things like heap overflows, and because the browser is not tested with asan/sanitizes, how confident will you say that there is no memory or other similar problems?

**Reply by csagan5** [reply]
> This is more like a general question, but I was actually thinking of something the other way around: what is the chance some code executions can cause a sandbox escape? When I was talking about the risks I was thinking about things like heap overflows, and because the browser is not tested with asan/sanitizes, how confident will you say that there is no memory or other similar problems?

It depends from the threat you are trying to model: Bromite patches almost never touch code that can be abused by a website to attack the visitor; only such patches would be vulnerable to what you mention. Luckily there are only a few and they are quite simple, so it is relatively easy to review them. I am referring to the patches touching the DOM document; part of them could be replaced by user scripts, eventually.

> the browser is not tested with asan/sanitizes

You mean the automated tests or? Bromite builds already use CFI (for non-x86 systems) and other compile-time hardening measures. If the tests you mention are running for Chromium then we should run them as well for Bromite.

**Reply by wchen342** [reply]
Yes I mean the automated tests. Do you run them for Bromite? Because I think the tests will not build properly after the patches.

**Reply by csagan5** [reply]
No, they are not currently being run; @uazo is looking into that as per previous message

**Comment by csagan5** [comment]
@wchen342 if you want to start merging some patches to Bromite please begin; also documentation and any useful information you would like to incorporate, it would be great to have it.

Currently @uazo is reviewing the patches for various issues/completeness; I can add also you as a collaborator so it's easier to organise the issues/PRs in future.

You should be able to easily find some of your patches and UC patches in Bromite because I preserved authorship (you can grep); there are minor changes due to porting over time and adjustments to not conflict with other Bromite patches. We can discuss everything in separate discussions/issues/PRs.

I am also thinking how to have some sort of (fast) C.I. with custom runners, but no promises yet about this.

The project needs all the help it can get, but of course take your time and don't overload yourself in any case, we should have fun, learn and make some accomplishment in these open source projects, nothing else.

**Reply by wchen342** [reply]
> I am also thinking how to have some sort of (fast) C.I. with custom runners

I am actually curious about what you are using currently? For `ungoogled-chromium-android` I just have actions triggered on tags and sent a request to a private repo where I have my signing keys as well as the Gandi server as a runner.

> take your time and don't overload yourself in any case

I certainly worked too much on the extensions for the past week, just cannot stand having something in hand that lies around unfinished. I am pretty sure I won't ever want to do that again after having experienced it once..

**Reply by csagan5** [reply]
> I am actually curious about what you are using currently?

A similar setup; do you keep the Gandi server up all the time?

**Reply by wchen342** [reply]
No, it will be too costful. They told me to shut it down when not in use from the beginning and I did so.

**Reply by PF4Public** [reply]
@wchen342 , @csagan5 Sorry for off-topic, but I wonder how did you both secure a server from Gandi? Did they offer it themselves or did you ask them for one?

**Reply by wchen342** [reply]
@PF4Public I asked them as a sponsor on behalf of Eloston.

**Reply by csagan5** [reply]
See https://github.com/ossfriendly/open-source-supporters

**Comment by r-a-y** [comment]
May I ask if there are any major differences when using the SystemWebView versions of Bromite vs. Ungoogled Chromium? The main benefit for Bromite is the ad filtering, but that isn't accessible when using the SystemWebView version (I believe, correct me if I'm wrong).

**Comment by BraINstinct0** [comment]
So, where are we now? I'm very curious about the Expected Time 'til this gets added to Bromite.

**Reply by wchen342** [reply]
I ended up doing it the other way around, so I added some of the Bromite stuff to ungoogled-chromium-android (mainly proxy, bookmark). The most important part of ungoogled-chromium-android, which is removing Google play integration as well as domain substitution is already in Bromite. The other parts are more like small improvements.

**Reply by csagan5** [reply]
> The other parts are more like small improvements.

I would be eager to accept these as patches, I just have not had time to go through all of them and compare.

Anyone who wants to do that is welcome (cc @uazo).

**Reply by PF4Public** [reply]
This should work both ways! :) I wonder if someone at ungoogled-software follows Bromite patches…

**Reply by csagan5** [reply]
@PF4Public you can do that for both projects and submit the PRs.

**Reply by uazo** [reply]
> Anyone who wants to do that is welcome (cc @uazo).

I had already looked at it (we took that of the auto update, ah, we never thanked him, thanks @wchen342!), but if @wchen342 indicates which ones may be interesting for bromite I double check.

**Reply by wchen342** [reply]
@uazo About that auto update thing, in last version Google removed the native part for Omaha so I am not sure it is actually working (I didn't have the time to actually test). Is your version in Bromite working?

**Reply by uazo** [reply]
> in last version Google removed the native part for Omaha

you are right, removed here
https://source.chromium.org/chromium/chromium/src/+/b0f8692e85340f0ddba4c46586401cce150bea9a
https://source.chromium.org/chromium/chromium/src/+/554073f89082a438f7b35f9f77c79a42a4c44b48

> Is your version in Bromite working?

when i did the patches for v96 i did not find any code that is no longer present.
I have to wait for the build to finish, then I'll check

**Reply by csagan5** [reply]
That means the current v96 does not auto-update? We might want to revert those changes and keep the inline updates code.

**Reply by uazo** [reply]
just checked, in v96 it still works.

**Reply by wchen342** [reply]
In that case I will just reverse adapt your patch I guess.

**Reply by csagan5** [reply]
It is still working probably because the native part is not used in Bromite.

**Reply by wchen342** [reply]
I tested my version and it still works, only lost the ability of setting intervals of checking updates.

**Reply by wchen342** [reply]
@uazo `InlineUpdateControllerFactory` is finally gone in cr97. Now almost all Omaha parts are removed on Java side so the updater won't work anymore. I am not sure what is the best way to keep the updater running now, maybe a cron-like job running periodically?

**Reply by uazo** [reply]
I can definitely take a look at it, just wait for the provisional v97 of bromite first, or @csagan5 you say that it is appropriate to work directly on the chromium vanilla code? because I think it is better not to go out with a v97 without autoupdate

> I am not sure what is the best way to keep the updater running now

the approach I would use is to check what we need of https://github.com/bromite/bromite/discussions/957#discussioncomment-1750227 trying to bring back as little as possible

**Reply by uazo** [reply]
@wchen342 here the patch for v97

https://github.com/bromite/bromite/blob/0a8ff322d6e9d738f7b2fa52618b178081bd100d/build/patches/Bromite-auto-updater.patch

as @csagan5 said it was not necessary to restore anything from the native.
it works, I tried it, however if you can check it too it would be better.

**Reply by wchen342** [reply]
@uazo Thanks, I'll try it and see whether it works here too.
